test/envtest/telemetryServer.go

```diff
@@ -0,0 +1,128 @@
+package envtest
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"io"
+	"net"
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/kong/kubernetes-ingress-controller/v3/test/helpers/certificate"
+)
+
+// TelemetryServer represents a server that listens for telemetry data over a TLS connection.
+type TelemetryServer struct {
+	reportChan chan []byte        // Channel to receive telemetry reports.
+	listener   net.Listener       // TLS listener for incoming telemetry connections.
+	started    bool               // Indicates whether the server is running.
+	cancel     context.CancelFunc // Function to cancel the server's context.
```

Comment:
This is OK but we generally try to add comments above the fields that they relate to.

This aligns with how Go does it.